File: insertion_sort_list.py
class ListNode:
    def __init__(self,x):
        self.val=x
        self.next=None
class Solution:
    def insertionSortList(self,head):
        if not head:
            return head
        dummy=ListNode(-1)
        dummy.next=head
        cur=head
        while cur.next:
            if cur.val>cur.next.val:
                pre=dummy
                while pre.next.val<cur.next.val:
                    pre=pre.next
                tmp=cur.next
                cur.next=tmp.next
                tmp.next=pre.next
                pre.next=tmp
            else:
                cur=cur.next
        return dummy.next
if __name__=='__main__':
    sol=Solution()
    List=head=ListNode(1)
    lst=[4,3,5,2]
    for x in lst:
        List.next=ListNode(x)
        List=List.next
    head1=sol.insertionSortList(head)
    #head1=sol.fuck(head)
    print 'fuck'
    while head1:
        print head1.val
        head1=head1.next

File: Algorithm/CPP/insertion_sort_list.cpp
# include <stdio.h>
# include <iostream>
# include <vector>
# include <algorithm>
using namespace std;


struct ListNode {
	int val;
	ListNode *next;
	ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
	ListNode* insertionSortList(ListNode* head) {
		if (!head || !head->next) return head;
		auto dummy = new ListNode(-1);
		dummy->next = head;
		auto p = head;
		auto q = head->next;
		while (q)
		{
			if (p->val > q->val)
			{
				p->next = q->next;
				auto tmp = dummy;
				while (tmp->next->val < q->val)
					tmp = tmp->next;
				q->next = tmp->next;
				tmp->next = q;
				q = p->next;
			}
			else
			{
				p = q;
				q = q->next;
			}
		}
		return dummy->next;
	}
};
int main()
{
	ListNode *head1 = new ListNode(1);
	ListNode *head2 = new ListNode(-2);
	auto p = head1;
	auto q = head2;
	vector<int> vec{ 4,3,2,5,6};//2
	for (auto x : vec)
	{
		p->next = new ListNode(x);
		p = p->next;
	}
	p->next = NULL;

	Solution sol;

	auto pp = sol.insertionSortList(head1);
	while (pp != NULL)
	{
		cout << pp->val << " ";
		pp = pp->next;
	}
	cout << endl;
	system("pause");
	return 0;
}
